### DISCUSSION TITLE
Failure in installing pynwb

### DISCUSSION THREAD

**Discussion by HarelShouval** [NeurodataWithoutBorders/helpdesk#40] [General]
I am trying to install pynwb in order to read .nwb files.
I used pip install -U pynwb

The install fails, and i get the following message:

"...
 Building h5py requires pkg-config unless the HDF5 path is explicitly specified using the environment variable HDF5_DIR. For more information and details, see https://docs.h5py.org/en/stable/build.html#custom-installation
      error: pkg-config probably not installed: FileNotFoundError(2, 'No such file or directory')
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for h5py
Failed to build h5py
ERROR: Could not build wheels for h5py, which is required to install pyproject.toml-based projects
"
It seems i am missing some file, but i have no idea how to fix this.

Thanks
Harel

**Comment by bendichter** [comment]
@HarelShouval what version of Python are you using? It looks like you may be using 3.11, which is not yet supported, since the h5py dependency isn't supported yet

**Reply by HarelShouval** [reply]
I am getting the same error when specifically using pip3.10

**Reply by rly** [reply]
We need a little more information to help diagnose the issue. What do you get when you run the following:
```
which python
python --version
which pip
```

**Comment by HarelShouval** [comment]
OK, I installed 3.10 as well thinking this is the problem, however I might need to unistall the 3.11, which I have not figured out how to do yet.

Thanks

From: Ben Dichter ***@***.***>
Date: Tuesday, November 29, 2022 at 11:25 AM
To: NeurodataWithoutBorders/helpdesk ***@***.***>
Cc: Shouval, Harel ***@***.***>, Mention ***@***.***>
Subject: Re: [NeurodataWithoutBorders/helpdesk] Failure in installing pynwb (Discussion #40)
External: Increase caution when handling links and attachments.


@HarelShouval<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FHarelShouval&data=05%7C01%7Charel.shouval%40uth.tmc.edu%7C3d81d3534fc14d25314808dad22ec538%7C7b326d2441ad4f57bc6089e4a6ac721b%7C0%7C0%7C638053395572031181%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=py8prwzSgcH1kAfkf98nOIdztXNDLp3jpG7s15NwxNM%3D&reserved=0> what version of Python are you using? It looks like you may be using 3.11, which is not yet supported, since the h5py dependency isn't supported yet

**Comment by bendichter** [comment]
I would recommend using [`conda`](https://conda.io/projects/conda/en/latest/user-guide/install/index.html). Then you can create a conda env using

```shell
conda create -n nwb python=3.9
conda activate nwb
```

**Comment by HarelShouval** [comment]
OK, the problem was indeed with 3.11
I uninstalled it and installed 3.9 and it worked.

Thanks much for your help

Harel



From: Ryan Ly ***@***.***>
Date: Wednesday, November 30, 2022 at 2:34 AM
To: NeurodataWithoutBorders/helpdesk ***@***.***>
Cc: Shouval, Harel ***@***.***>, Mention ***@***.***>
Subject: Re: [NeurodataWithoutBorders/helpdesk] Failure in installing pynwb (Discussion #40)
External: Increase caution when handling links and attachments.


We need a little more information to help diagnose the issue. What do you get when you run the following:

which python

python --version

which pip

**Comment by HarelShouval** [comment]
I am still having problems. With Python 3.9 I was able to install the nwb package.
However, when running code I still get an error message.
This is not my code, I am just using it to read other people’s data. Therefore, I know it is indeed working.

This is the error message I get:

Traceback (most recent call last):

  File "<stdin>", line 1, in <module>

  File "<string>", line 39, in <module>

  File "/Users/hshouval/Documents/Papers/Ian VTA/Huijeong dataset/photometry_python/functions/load.py", line 284, in load_nwb

    io = NWBHDF5IO(url, mode='r', driver='ros3')

  File "/Users/hshouval/opt/miniconda3/lib/python3.9/site-packages/hdmf/utils.py", line 645, in func_call

    return func(args[0], **pargs)

  File "/Users/hshouval/opt/miniconda3/lib/python3.9/site-packages/pynwb/__init__.py", line 234, in __init__

    super().__init__(path, manager=manager, mode=mode, file=file_obj, comm=comm, driver=driver)

  File "/Users/hshouval/opt/miniconda3/lib/python3.9/site-packages/hdmf/utils.py", line 645, in func_call

    return func(args[0], **pargs)

  File "/Users/hshouval/opt/miniconda3/lib/python3.9/site-packages/hdmf/backends/hdf5/h5tools.py", line 85, in __init__

    super().__init__(manager, source=path)

  File "/Users/hshouval/opt/miniconda3/lib/python3.9/site-packages/hdmf/utils.py", line 645, in func_call

    return func(args[0], **pargs)

  File "/Users/hshouval/opt/miniconda3/lib/python3.9/site-packages/hdmf/backends/io.py", line 23, in __init__

    self.open()

  File "/Users/hshouval/opt/miniconda3/lib/python3.9/site-packages/hdmf/backends/hdf5/h5tools.py", line 729, in open

    self.__file = File(self.source, open_flag, **kwargs)

  File "/Users/hshouval/opt/miniconda3/lib/python3.9/site-packages/h5py/_hl/files.py", line 485, in __init__

    raise ValueError(

ValueError: h5py was built without ROS3 support, can't use ros3 driver


My highlight.

Thanks
Harel

From: Ben Dichter ***@***.***>
Date: Tuesday, November 29, 2022 at 11:25 AM
To: NeurodataWithoutBorders/helpdesk ***@***.***>
Cc: Shouval, Harel ***@***.***>, Mention ***@***.***>
Subject: Re: [NeurodataWithoutBorders/helpdesk] Failure in installing pynwb (Discussion #40)
External: Increase caution when handling links and attachments.


@HarelShouval<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FHarelShouval&data=05%7C01%7Charel.shouval%40uth.tmc.edu%7C3d81d3534fc14d25314808dad22ec538%7C7b326d2441ad4f57bc6089e4a6ac721b%7C0%7C0%7C638053395572031181%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=py8prwzSgcH1kAfkf98nOIdztXNDLp3jpG7s15NwxNM%3D&reserved=0> what version of Python are you using? It looks like you may be using 3.11, which is not yet supported, since the h5py dependency isn't supported yet

**Reply by bendichter** [reply]
@HarelShouval it looks like you are using the ros3 driver for h5py. This is a possible way to read NWB data, but it is a pain to install h5py with ros3 enabled. A much easier (and faster!) approach is to use fsspec. See the documentation [here](https://pynwb.readthedocs.io/en/stable/tutorials/advanced_io/streaming.html#streaming-method-2-fsspec).

Try this:

```python
import fsspec
import pynwb
import h5py

# next, open the file
f = fs.open(s3_url, "rb")
file = h5py.File(f)
io = pynwb.NWBHDF5IO(file=file, load_namespaces=True)
nwbfile = io.read()

# do stuff

io.close()
file.close()
f.close()
```

**Comment by HarelShouval** [comment]
On Dec 1, 2022, at 10:06 AM, Ben Dichter ***@***.***> wrote:

﻿
External: Increase caution when handling links and attachments.


@HarelShouval<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FHarelShouval&data=05%7C01%7Charel.shouval%40uth.tmc.edu%7C921f15221225408d6e9808dad3b5f9c2%7C7b326d2441ad4f57bc6089e4a6ac721b%7C0%7C0%7C638055075776480992%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=Lvz1sinp0xYpoIp3wgWKwiv8TitfonQvNryJ7qWP%2Frg%3D&reserved=0> it looks like you are using the ros3 driver for h5py. This is a possible way to read NWB data, but it is a pain to install h5py with ros3 enabled. A much easier (and faster!) approach is to use fsspec. See the documentation here<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fpynwb.readthedocs.io%2Fen%2Fstable%2Ftutorials%2Fadvanced_io%2Fstreaming.html%23streaming-method-2-fsspec&data=05%7C01%7Charel.shouval%40uth.tmc.edu%7C921f15221225408d6e9808dad3b5f9c2%7C7b326d2441ad4f57bc6089e4a6ac721b%7C0%7C0%7C638055075776480992%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=IAcagwHyuvFHAIRMfqpjyWuqbGn0Fnd7QQdslVrHmdU%3D&reserved=0>.

Try this:

import fsspec
import pynwb
import h5py

# next, open the file
f = fs.open(s3_url, "rb")
file = h5py.File(f)
io = pynwb.NWBHDF5IO(file=file, load_namespaces=True)
nwbfile = io.read()

# do stuff

io.close()
file.close()
f.close()

**Comment by HarelShouval** [comment]
Thanks for the quick response
Harel

On Dec 1, 2022, at 11:14 AM, Shouval, Harel ***@***.***> wrote:

﻿


On Dec 1, 2022, at 10:06 AM, Ben Dichter ***@***.***> wrote:

﻿
External: Increase caution when handling links and attachments.


@HarelShouval<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FHarelShouval&data=05%7C01%7Charel.shouval%40uth.tmc.edu%7C921f15221225408d6e9808dad3b5f9c2%7C7b326d2441ad4f57bc6089e4a6ac721b%7C0%7C0%7C638055075776480992%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=Lvz1sinp0xYpoIp3wgWKwiv8TitfonQvNryJ7qWP%2Frg%3D&reserved=0> it looks like you are using the ros3 driver for h5py. This is a possible way to read NWB data, but it is a pain to install h5py with ros3 enabled. A much easier (and faster!) approach is to use fsspec. See the documentation here<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fpynwb.readthedocs.io%2Fen%2Fstable%2Ftutorials%2Fadvanced_io%2Fstreaming.html%23streaming-method-2-fsspec&data=05%7C01%7Charel.shouval%40uth.tmc.edu%7C921f15221225408d6e9808dad3b5f9c2%7C7b326d2441ad4f57bc6089e4a6ac721b%7C0%7C0%7C638055075776480992%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=IAcagwHyuvFHAIRMfqpjyWuqbGn0Fnd7QQdslVrHmdU%3D&reserved=0>.

Try this:

import fsspec
import pynwb
import h5py

# next, open the file
f = fs.open(s3_url, "rb")
file = h5py.File(f)
io = pynwb.NWBHDF5IO(file=file, load_namespaces=True)
nwbfile = io.read()

# do stuff

io.close()
file.close()
f.close()

**Comment by HarelShouval** [comment]
I can try to mimic, though I don’t really understand, which causes problems.


In the line: f = fs.open(s3_url, "rb")

I assume that fs.open is fsspec which I can solve by import fsspec as fs.

However, what is s3_url?



Thanks

Harel



From: Ben Dichter ***@***.***>
Date: Thursday, December 1, 2022 at 10:06 AM
To: NeurodataWithoutBorders/helpdesk ***@***.***>
Cc: Shouval, Harel ***@***.***>, Mention ***@***.***>
Subject: Re: [NeurodataWithoutBorders/helpdesk] Failure in installing pynwb (Discussion #40)
External: Increase caution when handling links and attachments.


@HarelShouval<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FHarelShouval&data=05%7C01%7Charel.shouval%40uth.tmc.edu%7C921f15221225408d6e9808dad3b5f9c2%7C7b326d2441ad4f57bc6089e4a6ac721b%7C0%7C0%7C638055075776480992%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=Lvz1sinp0xYpoIp3wgWKwiv8TitfonQvNryJ7qWP%2Frg%3D&reserved=0> it looks like you are using the ros3 driver for h5py. This is a possible way to read NWB data, but it is a pain to install h5py with ros3 enabled. A much easier (and faster!) approach is to use fsspec. See the documentation here<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fpynwb.readthedocs.io%2Fen%2Fstable%2Ftutorials%2Fadvanced_io%2Fstreaming.html%23streaming-method-2-fsspec&data=05%7C01%7Charel.shouval%40uth.tmc.edu%7C921f15221225408d6e9808dad3b5f9c2%7C7b326d2441ad4f57bc6089e4a6ac721b%7C0%7C0%7C638055075776480992%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=IAcagwHyuvFHAIRMfqpjyWuqbGn0Fnd7QQdslVrHmdU%3D&reserved=0>.

Try this:

import fsspec

import pynwb

import h5py



# next, open the file

f = fs.open(s3_url, "rb")

file = h5py.File(f)

io = pynwb.NWBHDF5IO(file=file, load_namespaces=True)

nwbfile = io.read()



# do stuff



io.close()

file.close()

f.close()

**Reply by bendichter** [reply]
See the documentation [here](https://pynwb.readthedocs.io/en/stable/tutorials/advanced_io/streaming.html#getting-the-location-of-the-file-on-dandi)
